ext/drreg/drreg.c

```diff
@@ -266,6 +465,62 @@ drreg_max_slots_used(OUT uint *max)
  * ANALYSIS AND CROSS-APP-INSTR
  */
 
+#ifdef X86
+/* Returns true if state has been set */
+static bool
+determine_simd_liveliness_state(instr_t *inst, reg_id_t reg, void **value)
+{
+    ASSERT(value != NULL, "cannot be NULL");
+    ASSERT(reg_is_vector_simd(reg), "must be a vector SIMD register");
+
+    /* Reason over partial registers in SIMD case to achieve efficient spilling*/
+    reg_id_t xmm_reg = reg_resize_to_opsz(reg, OPSZ_16);
+    reg_id_t ymm_reg = reg_resize_to_opsz(reg, OPSZ_32);
+    reg_id_t zmm_reg = reg_resize_to_opsz(reg, OPSZ_64);
+
+    /* It is important to give precedence to bigger registers.
+     * If both ZMM0 and YMM0 are read and therefore live, then
+     * SIMD_ZMM_LIVE must be assigned and not SIMD_YMM_LIVE.
+     *
+     * The inverse also needs to be maintained. If both
+     * ZMM0 and YMM0 are dead, then SIMD_ZMM_DEAD must be
+     * assigned and not SIMD_YMM_DEAD.
+     *
+     * This is important to achieve efficient spilling/restoring.
+     */
+    if (instr_reads_from_reg(inst, zmm_reg, DR_QUERY_INCLUDE_COND_SRCS)) {
+        if (instr_reads_from_exact_reg(inst, zmm_reg, DR_QUERY_INCLUDE_COND_SRCS) &&
+            *value <= SIMD_ZMM_LIVE) {
+            *value = SIMD_ZMM_LIVE;
+        } else if (instr_reads_from_exact_reg(inst, ymm_reg,
+                                              DR_QUERY_INCLUDE_COND_SRCS) &&
+                   *value <= SIMD_YMM_LIVE) {
+            *value = SIMD_YMM_LIVE;
+        } else if (instr_reads_from_exact_reg(inst, xmm_reg,
+                                              DR_QUERY_INCLUDE_COND_SRCS) &&
+                   *value <= SIMD_XMM_LIVE) {
+            *value = SIMD_XMM_LIVE;
+        }
```

Comment:
Have an else with an assert as a sanity check.  Ditto below.

Reply:
Done.

Reply:
This is a good catch. This is actually a bug. We don't account for partial register reads here. For example we could read with a movhpd from xmm[8 bytes], which isn't matched by the read condition, but live should be set.

We can't just assert on the else part for the same reason though. The *write* part needs to just return false in the `else` part as-is. The read part needs to be updated and differentiate between partially reading from xmm, ymm, or zmm.

Reply:
I added a test-case for this issue. Bug fix is still pending.

Reply:
Thanks (in advance)

Reply:
Fixed. The reason why tests were not failing because the liveness value is initialised to LIVE outside the function, hence explicitly handling partial writes. The fix now is mored direct and implicit.

Please hit resolve if you agree with the fix.

Reply:
This now means though that a partial read from a xmm will track it as a zmm live, which may cause it to spill a zmm (once they will be supported). This could cause drreg to introduce AVX-512 instructions in SSE code for example (again, only once we add support for spilling zmm). I think in the context of how DynamoRIo manages its register operands (doesn't always tell you exactly which bytes are touched, just how many), the best would be to check for partial_read_from_xmm, partial_read_from_ymm, or partial_read_from_zmm. I don't think we need API functions for this, maybe small local helpers are enough.

Reply:
@johnfxgalea were you planning to address this? Otherwise I might be able to look at it soon.

Reply:
Yes fixing now!

Reply:
Done

Reply:
Thanks for fixing!